### PR TITLE
JAVA-1587: Deterministic ordering of columns used in mapper query

### PR DESCRIPTION
**Pull request by collinsauve**
See https://datastax-oss.atlassian.net/browse/JAVA-1587

(P.S. this is my first time contributing to java-driver, sorry if I make any obvious mistakes)

### CODE REVIEW

**Comment by tolbertam** [comment]
Hi @collinsauve, at first glance this looks great.  Agree that it would be nice to consistently order the columns in the prepared statement.  Taking a deeper look and will provide some more feedback or my thumbs up later today.

**Comment by tolbertam** [comment]
Just wrote a quick test, that creates 5 Sessions and Mappers for a User object with fields `login`, `phone`, and `name`:

Before change:
```
INSERT INTO ks_1.user (phone,login,name) VALUES (?,?,?);
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
INSERT INTO ks_1.user (phone,name,login) VALUES (?,?,?);
INSERT INTO ks_1.user (name,phone,login) VALUES (?,?,?);
```

With change:
```
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
INSERT INTO ks_1.user (login,name,phone) VALUES (?,?,?);
```

So this is working great!  Would you mind if I threw a test that validates this on your branch?

File: driver-mapping/src/main/java/com/datastax/driver/mapping/Mapper.java
Comment (tolbertam) [inline driver-mapping/src/main/java/com/datastax/driver/mapping/Mapper.java]: Small nit, instead of creating a new comparator each time, you could just have 1 static comparator, i.e.:

```java
    private static final Comparator<AliasedMappedProperty> PROPERTY_COMPARATOR = new Comparator<AliasedMappedProperty>() {
        @Override
        public int compare(AliasedMappedProperty o1, AliasedMappedProperty o2) {
            String col1Name = o1.mappedProperty.getMappedName();
            String col2Name = o2.mappedProperty.getMappedName();
            return col1Name.compareTo(col2Name);
        }
    };
```

**Comment by tolbertam** [comment]
[Here's a patch](https://gist.github.com/tolbertam/0fa9918858f79fd278763fc7864c20b0) that adds a test for validating the fix.  If it's ok with you I can add it to this branch.  Other than [my comment about making the comparator static](https://github.com/datastax/java-driver/pull/864#discussion_r130642148), this looks great to me 👍

**Comment by collinsauve** [comment]
Those changes sound good to me, do I need to give you write access to my fork to let you modify the PR?

**Comment by tolbertam** [comment]
I believe I have access, will try it out, thanks!

**Comment by tolbertam** [comment]
Pushed the test and the changes.  After thinking about it more, I figured we might as well make `AliasedMappedProperty` `Comparable` which accomplishes the same thing.

I also just realized we weren't considering `alias`, but I don't think that will really be a problem unless you have the same column/selector twice and you want to differentiate with an alias,  I think while possible that is not practical and not sure if it is worth adding it to compareTo (but will leave that to the other devs to judge when they review).

**Review by tolbertam — APPROVED** [review]

**Review by adutra — APPROVED** [review]
👍 one small nit though: changelog entry is missing.

**Review by olim7t — APPROVED** [review]

**Comment by DukeDai** [comment]
The issue for query is still there, 3.4 and 3.6. Considering that cassandra issue https://issues.apache.org/jira/browse/CASSANDRA-13641. Not sure whether later version has fixed similar issue for query or not. Could you please confirm?

**Comment by tolbertam** [comment]
the issue should be fixed from driver 3.3.1+, can you share a test case where you observe that this isn't the case?

**Pull request merged**
